h2o-core/src/main/java/water/rapids/ast/prims/advmath/AstStratifiedSplit.java

```diff
@@ -45,27 +52,75 @@ public ValFrame apply(Env env, Env.StackHelp stk, AstRoot asts[]) {
     seed = seed == -1 ? new Random().nextLong() : seed;
     final long[] classes = new VecUtils.CollectDomain().doAll(y).domain();
     final int nClass = y.isNumeric() ? classes.length : y.domain().length;
+    final String[] domains = y.domain();
     final long[] seeds = new long[nClass]; // seed for each regular fold column (one per class)
     for (int i = 0; i < nClass; ++i)
       seeds[i] = getRNG(seed + i).nextLong();
     String[] dom = new String[]{"train", "test"};
-    return new ValFrame(new MRTask() {
-      private boolean isTest(int row, long seed) {
-        return getRNG(row + seed).nextDouble() <= testFrac;
-      }
+    // create frame with all 0s (default is train)
+    Key<Frame> k1 = Key.make();
+    Vec resVec = Vec.makeCon(0,fr.anyVec().length());
+    resVec.setDomain(new String[]{"train","test"});
+    Frame result = new Frame(k1, new String[]{"test_train_split"}, new Vec[]{resVec});
+    DKV.put(result);
+    // create index frame
+    Key<Frame> k2 = Key.make();
+    Frame ones = new Frame(k2, new String[]{"ones"}, new Vec[]{Vec.makeCon(1.0,fr.anyVec().length())});
+    DKV.put(ones);
+    Frame idx = Rapids.exec("(cumsum " + ones._key + " 0)").getFrame();
+    DKV.put(idx);
+    // loop through each class
+    for (int classLabel = 0; classLabel < nClass; ++classLabel) {
```

Comment:
Im rethinking this strategy. The slowness comes from scanning the frame for every class. Would be faster to do one MRTask that generates a HashMap<Int,HashSet<Int>>